File: DEC2023/binary_search.py
# def binary_search(X,l,r,key):
#     if l>r:
#         return -1
#     mid = l + (r-l)//2
#     # print(mid,X[mid])
#     if X[mid] == key:
#         return mid
#     if X[mid]>key:
#         return binary_search(X,l,mid-1,key)
#     else:
#         return binary_search(X,mid+1,r,key)
def binary_search(arr,l,r,target):
    
    while l<=r:
        mid = l + (r-l)//2
        if arr[mid] == target:
            return mid
        elif arr[mid]>target:
            r = mid-1
        else:
            l = mid+1
    return -1


arr = [1,2,3,4,5]
n = len(arr)-1
print(binary_search(arr,0,n,4)) 

File: AUG2026/Longest Increasing Subsequence/binary-search.java
import java.util.*;

class Solution {
    public static void main(String args[]) {
        int numbers[] = {0,1,0,3,2,3};

        System.out.println(longestIncreasingSubsequence(numbers));
    }

    public static int longestIncreasingSubsequence(int[] numbers) {
        int N  = numbers.length;
        
        ArrayList<Integer> sub = new ArrayList<>();
        sub.add(numbers[0])
;
        for (int i = 1; i < N; i++) {
            if (numbers[i] > sub.get(sub.size() - 1)) {
                sub.add(numbers[i]);
            } else {
                int pos = binarySearch(sub, numbers[i]);

                sub.set(pos, numbers[i]);
            }
        }

        return sub.size();
    }

    public static int binarySearch(ArrayList<Integer> sub, int K) {
        int lo = 0,
            hi = sub.size() - 1;

        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;

            if (sub.get(mid) < K) {
                lo = mid + 1;
            } else {
                // Greater or equal
                hi = mid;
            }
        }

        return lo;
    }
}
